File: Leetcode/3-Longest-Substring-Without-Repeating-Characters/3-Longest-Substring-Without-Repeating-Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        memo = set([])
        i = 0
        j = 0
        res = 0
        curr = 0
        lenS = len(s)
        while j < lenS:
            if s[j] not in memo:
                memo.add(s[j])
                curr += 1
                j += 1
                res = max(res, curr)
            else:
                memo.remove(s[i])
                i += 1
                curr -= 1
        return res

File: Leetcode/3-Longest-Substring-Without-Repeating-Characters/3-Longest-Substring-Without-Repeating-Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0, right = 0, maxLen = 0;
        HashSet<Character> seen = new HashSet<Character>();

        while (left <= right && right < s.length()) {
            if (!seen.contains(s.charAt(right))) {
                seen.add(s.charAt(right));
                maxLen = Math.max(maxLen, right - left + 1);
                right++;
            } else {
                seen.remove(s.charAt(left));
                left++;
            }
        }

        return maxLen;
    }
}